### ISSUE TITLE
False negative no-member when using `hasattr`

### ISSUE THREAD

**Issue by allanlewis** [PyCQA/pylint#801]
I have some code like this:

``` py
if hasattr(foo, 'bar'):
    foo.bar = 'baz'
```

`foo` is an instance of a class, some of whose subclasses have an attribute `bar` and some don't. Pylint raises `no-member` for this because `foo` doesn't have a member `bar`. Irrespective of the merits or otherwise of this approach ("permission vs forgiveness"), surely Pylint shouldn't complain about this because I've already checked using `hasattr`? (I'm also aware of the issues around `hasattr`, but given it's a built-in I think it's reasonable to Pylint to assume it works.)

**Comment by PCManticore** [comment]
The problem is that currently pylint doesn't have flow sensitive inference, in the sense that will not consider the if guard when checking that bar is not defined. This is something that we want to introduce at some point though.

But the main problem with your example though is that pylint shouldn't trigger at all no-member for assigning attributes and I can't actually reproduce it. Maybe your original code is a bit different?

**Comment by rogalski** [comment]
Needs `control_flow` label.

**Comment by brycepg** [comment]
I think this is dooable without a control flow graph (same way we ignore exceptions)

Here's a reproduction:

    class Foof():
        pass
    foof = Foof()
    if hasattr(foof, 'bar'):
        foof.bar()

**Comment by allanlewis** [comment]
Closing as stale.

**Comment by Pierre-Sassoulas** [comment]
Please don't close issues simply because they are stall, it's just going to create more work when they get opened again and have to be triaged again. This issue has 5 thumbs-ups, so it might be interesting enough for someone to fix it.

**Comment by jacobtylerwalls** [comment]
Fixed in 107f59e949f1824251d14717eb57650273c0ffff

**Comment by Pierre-Sassoulas** [comment]
Thank you for the triaging @jacobtylerwalls, much appreciated.
